plan/planbuilder.go

```diff
@@ -328,6 +332,101 @@ func (b *planBuilder) buildAdmin(as *ast.AdminStmt) Plan {
 	return p
 }
 
+// getColumnOffsets returns the offsets of index columns, normal columns and primary key with integer type.
+func getColumnOffsets(tn *ast.TableName) (indexOffsets []int, columnOffsets []int, pkOffset int) {
+	tbl := tn.TableInfo
+	// idxNames contains all the normal columns that can be analyzed more effectively, because those columns occur as index
+	// columns or primary key columns with integer type.
+	var idxNames []string
+	pkOffset = -1
+	if tbl.PKIsHandle {
+		for i, col := range tbl.Columns {
+			if mysql.HasPriKeyFlag(col.Flag) {
+				idxNames = append(idxNames, col.Name.L)
+				pkOffset = i
+			}
+		}
+	}
+	indices, _ := availableIndices(tn.IndexHints, tn.TableInfo)
+	for _, index := range indices {
+		for i, idx := range tn.TableInfo.Indices {
+			if index.Name.L == idx.Name.L {
+				indexOffsets = append(indexOffsets, i)
+				break
+			}
+		}
+		if len(index.Columns) == 1 {
+			idxNames = append(idxNames, index.Columns[0].Name.L)
+		}
+	}
+	for i, col := range tbl.Columns {
+		isIndexCol := false
+		for _, idx := range idxNames {
+			if idx == col.Name.L {
+				isIndexCol = true
+				break
+			}
+		}
+		if !isIndexCol {
+			columnOffsets = append(columnOffsets, i)
+		}
+	}
+	return
+}
+
+func (b *planBuilder) prepareSimpleSelect(colNames, tableName string, ordered bool) Plan {
```

Comment:
I think you don't need to construct a raw sql stmt. Which may introduce redundent `Projection`. You can add a DataSource as `Analyze`'s child. Because in physical plan builder you also add Physical Index Scan manually.